Add release-branch-semver scheme

chrisjbillington opened the pull request:
This is a scheme that assumes that the upcoming release for most branches is a minor release, and that only for release branches is the upcoming version a patch release. This matches the workflow used by many projects who do development in master with feature branches, and fork off release branches that have bugfix-only changes made to them.

The scheme only considers the current branch to be a release branch if its name (stripped of any namespaces) parses as a version that matches the most recent tag up to the minor segment. Thus a tag `1.2.0rc0`for example will be matched by `1.2`, `1.2.x`, `maintenance/1.2`, `release-1.2.x` etc, but not `bugfix/123123`, since although the latter parses as a `LegacyVersion` `123123.0.0`, it doesn't match the tag up to the minor segment. So that's intended to reduce false positives from branch names that happen to parse as versions even though they're not.

This PR also increases the level of detail documenting the differences between the schemes in `README.rst`

chrisjbillington committed:
Add release-branch-semver scheme

This is a scheme that assumes that the upcoming release for most
branches is a minor release, and that only for release branches
(those whose branch names parse as a version number according to current
configuration) is the upcoming version a patch release.

Better document the differences between the schemes in README.rst

RonnyPfannschmidt commented on src/setuptools_scm/version.py:
```diff
@@ -132,6 +132,7 @@ def __init__(
         dirty=False,
         preformatted=False,
         branch=None,
+        config=None,
```
i am on the fence on whether configuration should be a parameter to versions

chrisjbillington replied:
That's fair enough, but if the branch name is to be parsed it needs config from somewhere - I was having it use the default config, but this way it can honour version regexes set by the user, I guess. Happy to have it just use the default config though!

RonnyPfannschmidt replied:
i understand the necessity - i'm just wondering about how to pass in the config
it must land in the version formatter/generator somehow

and its also going to be useful for a semver-towncrier scheme, that will pick versions number based on 

i'll sleep a night over this

both the pass in as part of the version object as well as adding it as extra parameter to the version formatter have different ups and downs

RonnyPfannschmidt reviewed (commented):
thanks for bringing this addition to the table, it makes a few details much more nice (and actually matches something i want/need for pytest anyway

chrisjbillington committed:
Allow guess_next_simple_semver to ignore unneeded parts

guess_next_simple_semver previously would error on trying to convert the
'rc0' part of a version such as '1.2.0rc1' to an int, even if it was
told to retain only the first two parts. (It would actually only error
on Python 2, where the conversion to int was non-lazy)

Now if it's told to retain only the first two parts, it won't try to
convert anything after the first two dots to an int, so it won't error.
The outcome is the same as if the bit after the second dot was a valid
int (which was going to be ignored anyway, we just now don't error on
it).

This is useful for the release-branch-semver scheme, which wants to be
able to find the next minor version even if the most recent version is a
release candidate.

RonnyPfannschmidt commented:
bringing this in as is for now as the other way around is potentially problematic as well and it wil lhelp the use-cases for towncrier based versioning as well

chrisjbillington commented:
Thank you!
